tensorboard/plugins/projector/vz_projector/knn.ts

```diff
@@ -59,47 +59,41 @@ export function findKNNGPUCosine<T>(
   // A*A.T will give us NxN matrix holding the cosine distance between every
   // pair of points, which we sort using KMin data structure to obtain the
   // K nearest neighbors for each point.
-  let typedArray = vector.toTypedArray(dataPoints, accessor);
-  const bigMatrix = tf.tensor(typedArray, [N, dim]);
-  let nearest: NearestEntry[][] = new Array(N);
-  let numPieces = Math.ceil(N / OPTIMAL_GPU_BLOCK_SIZE);
+  const nearest: NearestEntry[][] = new Array(N);
+  const numPieces = Math.ceil(N / OPTIMAL_GPU_BLOCK_SIZE);
   let M = Math.floor(N / numPieces);
   let modulo = N % numPieces;
   let offset = 0;
   let progress = 0;
   let progressDiff = 1 / (2 * numPieces);
   let piece = 0;
+
+  const typedArray = vector.toTypedArray(dataPoints, accessor);
+  const bigMatrix = tf.tensor(typedArray, [N, dim]);
+  const bigMatrixTransposed = bigMatrix.transpose();
+  const bigMatrixSquared = tf.matMul(bigMatrix, bigMatrixTransposed);
+  const cosDist = tf.sub(1, bigMatrixSquared);
+  const splits = tf.split(cosDist, numPieces, 1);
+
   function step(resolve: (result: NearestEntry[][]) => void) {
     let progressMsg =
       'Finding nearest neighbors: ' + (progress * 100).toFixed() + '%';
     util
       .runAsyncTask(
         progressMsg,
         async () => {
-          let B = piece < modulo ? M + 1 : M;
-          let typedB = new Float32Array(B * dim);
-          for (let i = 0; i < B; ++i) {
-            let vector = accessor(dataPoints[offset + i]);
-            for (let d = 0; d < dim; ++d) {
-              typedB[i * dim + d] = vector[d];
-            }
-          }
-          const partialMatrix = tf.tensor(typedB, [dim, B]);
-
-          const result = tf.matMul(bigMatrix, partialMatrix);
-          const partial = await result.array();
-          partialMatrix.dispose();
-          result.dispose();
-
+          const B = piece < modulo ? M + 1 : M;
+          const partial = await splits[piece].data();
           progress += progressDiff;
           for (let i = 0; i < B; i++) {
             let kMin = new KMin<NearestEntry>(k);
             let iReal = offset + i;
             for (let j = 0; j < N; j++) {
+              // Skip diagonal entries.
               if (j === iReal) {
                 continue;
               }
-              let cosDist = 1 - partial[j][i];
+              const cosDist = partial[j * B + i];
```

Comment:
I'm a bit confused.  iiuc, the shapes go like:
bigMatrix ~ [N, dim]
cosDist (on L75) ~ [N, N]
splits: partials[]
partial ~ [N, 256 /* OPTIMAL_GPU_BLOCK_SIZE */]

If `partial` has length N, then indexing it on `j * B + i` doesn't make sense to me here.  `j` can go up to N, `B` can go up to `M`, and `i` can go up to `M`, but it doesn't make sense for us to access any row in `partial` beyond `N`.

Could you help me understand what this formula is doing?

Reply:
Partial <- Float32Array (not double array) of `N * N / split_size`.

Reply:
Ahh, gotcha.  The added comment helpfully indicates how `.data()` causes reshaping, thanks.